### PR TITLE
Simplify `toString` representation of `State`

### PR DESCRIPTION
**Pull request by ebabel**
- Adds a cleaned up `toString` implementation with no package names. 
- Removed the hopefully redundant _"status="_ for both `Disconnected` data class and `Unknown` data class🤞 but can add back if preferred.
- Used non-reflective approach with `when`.
- Showing off the results in the included test.
- Renamed existing `StateTest` for clarity. 🤷

### CODE REVIEW

File: core/src/commonMain/kotlin/State.kt [hunk 1]
```diff
@@ -104,7 +104,42 @@ public sealed class State {
             public object EncryptionTimedOut : Status()
 
             /** Catch-all for any statuses that are unknown for a platform. */
-            public data class Unknown(val status: Int) : Status()
+            public data class Unknown(val status: Int) : Status() {
+                override fun toString(): String {
+                    return status.toString()
+                }
+            }
```
Comment (twyatt) [inline core/src/commonMain/kotlin/State.kt:111]: Please revert this change as I think it would make sense for consumers to be able to print an `Unknown` instance and have it take the form of `"Unknown(status=#)"`.

See comment below on `Disconnected.toString` where it handles the abbreviated `Unknown` printout.

File: core/src/commonMain/kotlin/State.kt [hunk 2]
```diff
@@ -104,7 +104,42 @@ public sealed class State {
...
+                    ConnectionLimitReached -> "Connection Limit Reached"
+                    EncryptionTimedOut -> "Encryption Timed Out"
+                    is Unknown -> "" // check override
+                }
+            }
+        }
+
+        override fun toString(): String {
+            return "Disconnected(${status.toString()})"
+        }
```
Comment (twyatt) [inline core/src/commonMain/kotlin/State.kt:132]:
```suggestion
        override fun toString(): String =
            "Disconnected(${if (status is Status.Unknown) status.status else status.toString()})"
```

File: core/src/commonMain/kotlin/State.kt [hunk 3]
```diff
@@ -104,7 +104,42 @@ public sealed class State {
...
+                    L2CapFailure -> "L2Cap Failure"
+                    Timeout -> "Timeout"
+                    LinkManagerProtocolTimeout -> "LinkManager Protocol Timeout"
+                    UnknownDevice -> "Unknown Device"
+                    Cancelled -> "Cancelled"
+                    ConnectionLimitReached -> "Connection Limit Reached"
+                    EncryptionTimedOut -> "Encryption Timed Out"
+                    is Unknown -> "" // check override
+                }
+            }
```
Comment (twyatt) [inline core/src/commonMain/kotlin/State.kt:127]: Slightly more condensed as expression body syntax:

```suggestion
            override fun toString(): String = when (this) {
                PeripheralDisconnected -> "Peripheral Disconnected"
                CentralDisconnected -> "Central Disconnected"
                Failed -> "Failed"
                L2CapFailure -> "L2Cap Failure"
                Timeout -> "Timeout"
                LinkManagerProtocolTimeout -> "LinkManager Protocol Timeout"
                UnknownDevice -> "Unknown Device"
                Cancelled -> "Cancelled"
                ConnectionLimitReached -> "Connection Limit Reached"
                EncryptionTimedOut -> "Encryption Timed Out"
                is Unknown -> "" // check override
            }
```

File: core/src/commonTest/kotlin/StateStringTests.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        test("Disconnected(L2Cap Failure)", Disconnected(L2CapFailure))
+        test("Disconnected(Timeout)", Disconnected(Timeout))
+        test("Disconnected(LinkManager Protocol Timeout)", Disconnected(LinkManagerProtocolTimeout))
+        test("Disconnected(Unknown Device)", Disconnected(UnknownDevice))
+        test("Disconnected(Cancelled)", Disconnected(Cancelled))
+        test("Disconnected(Connection Limit Reached)", Disconnected(ConnectionLimitReached))
+        test("Disconnected(Encryption Timed Out)", Disconnected(EncryptionTimedOut))
+        test("Disconnected(133)", Disconnected(Unknown(133)))
+    }
+}
```
Comment (twyatt) [inline core/src/commonTest/kotlin/StateStringTests.kt:48]: Perhaps easier/nicer to read as?:

```suggestion
    @Test
    fun testToString() {
        Connected assertToStringEquals "Connected"
        Connected assertToStringEquals "Connected"
        Bluetooth assertToStringEquals "Connecting.Bluetooth"
        Observes assertToStringEquals "Connecting.Observes"
        Services assertToStringEquals "Connecting.Services"
        Disconnecting assertToStringEquals "Disconnecting"
        Disconnected(PeripheralDisconnected) assertToStringEquals "Disconnected(Peripheral Disconnected)"
        Disconnected(CentralDisconnected) assertToStringEquals "Disconnected(Central Disconnected)"
        Disconnected(Failed) assertToStringEquals "Disconnected(Failed)"
        Disconnected(L2CapFailure) assertToStringEquals "Disconnected(L2Cap Failure)"
        Disconnected(Timeout) assertToStringEquals "Disconnected(Timeout)"
        Disconnected(LinkManagerProtocolTimeout) assertToStringEquals "Disconnected(LinkManager Protocol Timeout)"
        Disconnected(UnknownDevice) assertToStringEquals "Disconnected(Unknown Device)"
        Disconnected(Cancelled) assertToStringEquals "Disconnected(Cancelled)"
        Disconnected(ConnectionLimitReached) assertToStringEquals "Disconnected(Connection Limit Reached)"
        Disconnected(EncryptionTimedOut) assertToStringEquals "Disconnected(Encryption Timed Out)"
        Disconnected(Unknown(133)) assertToStringEquals "Disconnected(133)"
    }
}

private infix fun State.assertToStringEquals(expected: String) {
    assertEquals(
        expected = expected,
        actual = toString(),
    )
}
```

File: core/src/commonTest/kotlin/StateIsActiveTests.kt [hunk 5]
```diff
@@ -4,7 +4,7 @@ import kotlin.test.Test
 import kotlin.test.assertFalse
 import kotlin.test.assertTrue
 
-public class StateTests {
+public class StateIsActiveTests {
```
Comment (twyatt) [inline core/src/commonTest/kotlin/StateIsActiveTests.kt:7]:
```suggestion
public class StateIsAtLeastTests {
```

File: core/src/commonMain/kotlin/State.kt [hunk 6]
```diff
@@ -104,7 +104,42 @@ public sealed class State {
...
+    }
+
+    override fun toString(): String {
+        return when (this) {
+            Connected -> "Connected"
+            Connecting.Bluetooth -> "Connecting.Bluetooth"
+            Connecting.Observes -> "Connecting.Observes"
+            Connecting.Services -> "Connecting.Services"
+            Disconnecting -> "Disconnecting"
+            is Disconnected -> "" // check override
```
Comment (twyatt) [inline core/src/commonMain/kotlin/State.kt:142]: Should either `error` or have a valid string. If a future change causes us to actually hit this item, then it would be bad to provide an empty string.

What does "check override" comment mean?

File: core/src/commonMain/kotlin/State.kt [hunk 7]
```diff
@@ -104,7 +104,42 @@ public sealed class State {
...
+            }
+        }
+
+        override fun toString(): String {
+            return "Disconnected(${status.toString()})"
+        }
+    }
+
+    override fun toString(): String {
+        return when (this) {
```
Comment (twyatt) [inline core/src/commonMain/kotlin/State.kt:136]: Expression body syntax, please.

File: core/src/commonMain/kotlin/State.kt [hunk 8]
```diff
@@ -104,7 +104,42 @@ public sealed class State {
...
+                    CentralDisconnected -> "Central Disconnected"
+                    Failed -> "Failed"
+                    L2CapFailure -> "L2Cap Failure"
+                    Timeout -> "Timeout"
+                    LinkManagerProtocolTimeout -> "LinkManager Protocol Timeout"
+                    UnknownDevice -> "Unknown Device"
+                    Cancelled -> "Cancelled"
+                    ConnectionLimitReached -> "Connection Limit Reached"
+                    EncryptionTimedOut -> "Encryption Timed Out"
+                    is Unknown -> "" // check override
```
Comment (twyatt) [inline core/src/commonMain/kotlin/State.kt:125]: Should either `error` or have a valid string. If a future change causes us to actually hit this item, then it would be bad to provide an empty string.

What does "check override" comment mean?

**Review by twyatt — CHANGES_REQUESTED** [review]
Very nice.

File: core/src/commonTest/kotlin/StateStringTests.kt (hunk 4 above)
Comment (ebabel) [reply core/src/commonTest/kotlin/StateStringTests.kt:48]: Unfortunately longer than the formatter allows for one line. Does that change your preference?

Comment (twyatt) [reply core/src/commonTest/kotlin/StateStringTests.kt:48]: `private` to this file, we can abbreviate further to `assertEquals` if that would fit?

**Comment by twyatt** [comment]
@ebabel `./gradlew apiDump`

File: core/src/commonTest/kotlin/StateStringTests.kt [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import com.juul.kable.State.Disconnected.Status.ConnectionLimitReached
+import com.juul.kable.State.Disconnected.Status.Unknown
+import com.juul.kable.State.Disconnecting
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+public class StateStringTests {
+
+    @Test
+    fun testToString() {
```
Comment (twyatt) [inline core/src/commonTest/kotlin/StateStringTests.kt:26]: These are tests, so `test` prefix is redundant.

```suggestion
    fun toString() {
```

Or (if `toString` is reserved):

```suggestion
    fun stringRepresentation() {
```

File: core/src/commonTest/kotlin/StateStringTests.kt [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        Disconnected(UnknownDevice).assertEquals("Disconnected(Unknown Device)")
+        Disconnected(Cancelled).assertEquals("Disconnected(Cancelled)")
+        Disconnected(ConnectionLimitReached).assertEquals("Disconnected(Connection Limit Reached)")
+        Disconnected(EncryptionTimedOut).assertEquals("Disconnected(Encryption Timed Out)")
+        Disconnected(Unknown(133)).assertEquals("Disconnected(133)")
+        Unknown(133).assertEquals("Unknown(status=133)")
+    }
+}
+
+private infix fun Any.assertEquals(expected: String) {
```
Comment (twyatt) [inline core/src/commonTest/kotlin/StateStringTests.kt:47]:
```suggestion
private fun State.assertEquals(expected: String) {
```

**Review by twyatt — APPROVED** [review]

**Review by davidtaylor-juul — APPROVED** [review]

**Review by cedrickcooke — APPROVED** [review]

**Comment by twyatt** [comment]
@ebabel Thanks for the contribution! It will be included in the next ([0.18.1](https://github.com/JuulLabs/kable/milestone/28)) release.
